include/fiction/algorithms/simulation/sidb/quickexact.hpp

```diff
@@ -0,0 +1,477 @@
+            {
+                charge_distribution_surface<Lyt> charge_lyt_copy{charge_lyt};
+                for (const auto& cell : detected_negative_sidbs)
+                {
+                    charge_lyt.adding_sidb_to_layout(cell, -1);
+                }
+                simulation_result.charge_distributions.push_back(charge_lyt_copy);
+            }
+        }
+        simulation_result.simulation_runtime = time_counter;
+        result                               = simulation_result;
+    }
+
+    sidb_simulation_result<Lyt> get_simulation_results() const
+    {
+        return result;
+    }
+
+  private:
+    /**
+     * This function conducts 2-state physical simulation (negative, neutral).
+     *
+     * @param charge_lyt_new Initialized charge layout.
+     * @param sim_result sidb_simulation_result to collect gained results.
+     */
+    void two_state_simulation(charge_distribution_surface<Lyt>& charge_lyt_new, sidb_simulation_result<Lyt>& sim_result)
+    {
+        charge_lyt_new.set_base_number(2);
+        uint64_t current_charge_index  = 0;
+        uint64_t previous_charge_index = 0;
+        for (uint64_t i = 0; i <= charge_lyt_new.get_max_charge_index(); i++)
+        {
+            current_charge_index = (i ^ (i >> 1));  // gray code is used for the charge index.
+            charge_lyt_new.set_charge_index_by_gray_code(
+                current_charge_index, previous_charge_index, false, false,
+                true);  // "false" allows that the charge state of the dependent cell is
+                        // automatically changed based on the new charge distribution.
+            previous_charge_index = current_charge_index;
+
+            if (charge_lyt_new.is_physically_valid())
+            {
+                charge_distribution_surface<Lyt> charge_lyt_copy{charge_lyt_new};
+                charge_lyt_copy.recompute_system_energy();
+                // The previously detected negatively charged SiDBs are added to the final layout.
+                for (const auto& cell : detected_negative_sidbs)
+                {
+                    charge_lyt_copy.adding_sidb_to_layout(cell, -1);
+                }
+                sim_result.charge_distributions.push_back(charge_lyt_copy);
+            }
+        }
+
+        // The cells of the previously detected negatively charged SiDBs are added to the cell level layout.
+        for (const auto& cell : detected_negative_sidbs)
+        {
+            layout.assign_cell_type(cell, Lyt::cell_type::NORMAL);
+        }
+    }
+    /**
+     * This function conducts 3-state physical simulation (negative, neutral, positive).
+     *
+     * @param charge_lyt_new Initialized charge layout.
+     * @param sim_result sidb_simulation_result to collect gained results.
+     */
+    void three_state_simulation(charge_distribution_surface<Lyt>& charge_lyt_new,
+                                sidb_simulation_result<Lyt>&      sim_result)
```

Comment:
No gray code here?

Reply:
Yes, there are several reasons. One of them is that I discovered that the 3-state simulation needs a new physical model because the current version of the physical validity is not powerful enough for 3-state simulation in general. So optimizing the 3-state simulation is something for later. However, this 3-state simulation is already very powerful, since only SiDBs that could be positively charged are simulated with the 3-state simulation.